### ISSUE TITLE
Support gradle project dependencies - to enable Android Studio library modules as java subprojects

### ISSUE THREAD

**Issue by buzzware** [HipByte/motion-gradle#6]
In Android Studio it's quite simple to create library modules :
1) Create a new project in AS
2) Create a library New Module -> Android Library
3) I had to right-click on the new subfolder and select "Make Module"
4) add the following to the app build.gradle :
dependencies {
    ...
    compile project(":lib")   // if your library is called lib
}

It would be a massive gain to make it this easy to add the library module above as a project dependency to a RubyMotion project. Without knowing the details of RubyMotion, it may simple because gradle is already set up for compiling an aar for the library. Its just a matter of the RubyMotion app doing the same as the Java app above (managing dependencies, compiling, linking).

By the way, I don't think a library module aar contains its dependencies - the app must compile and link them specifically, and so the RubyMotion project would need to do the same.

https://developer.android.com/sdk/installing/create-project.html
https://developer.android.com/tools/projects/projects-cmdline.html

Enabling Java subprojects would mean that the full Android Studio tool set and API is available for low level Java coding without the current limitations of RubyMotion. The result could be easily included and automatically compiled in to a RubyMotion project. Temporary bugs and limitations in RM could be overcome by falling back to Java, and the result can be translated to Ruby later.

**Comment by jjaffeux** [comment]
@buzzware Yes I think it should work, almost out of the box :)
I will look at it later, if you want to help me please build a RubyMotion project with your module in vendor folder, and a method in main activity using this module like : MyModule.test which should print 'test'

I will just have to plug `motion-gradle` after and try to make it work :) it would save me the setup time and I would be sure to understand what you need.

**Comment by jjaffeux** [comment]
Concerning your second point and dependencies in aar, I'm pretty sure you can have a gradle file in this lib, and motion-gradle should be able to read it and pull the dependencies from it.

**Comment by buzzware** [comment]
OK, I'm building an Android Studio app with module for you, with an embedded RubyMotion project. I know that's the inverse of what you asked for, but it should work as well and is easier for me as I've already done it (just with my app specific stuff). Please note I'm a beginner with Java, I'm relying on 20+ years with compilers in other languages.

**Comment by buzzware** [comment]
Here it is https://github.com/buzzware/MotionStudio

Basically the MotionStudioRuby folder project should be able to mimic the java app folder project. Thanks for getting onto this! I'm calling it a night here (11:30pm in Perth, Australia :))

**Comment by jjaffeux** [comment]
@buzzware 

➜  MotionStudioRuby git:(master) rake                                                                                                                   (05-21 17:54)
   Compile ./app/main_activity.rb
    Create ./build/Development-21/lib/armeabi/libpayload.so
   Compile ./app/main_activity.rb
    Create ./build/Development-21/lib/x86/libpayload.so
    Create ./build/Development-21/MotionStudioRuby.apk
      Sign ./build/Development-21/MotionStudioRuby.apk
     Align ./build/Development-21/MotionStudioRuby.apk
   Install ./build/Development-21/MotionStudioRuby.apk
1957 KB/s (3592190 bytes in 1.792s)
     Start com.yourcompany.motionstudioruby/.MainActivity
--------- beginning of main
--------- beginning of system
I/com/yourcompany/motionstudioruby( 1724): "test"

**Comment by buzzware** [comment]
Oh wow, veerrry cool!
I'll get onto doing this in my real app tomorrow, but can I please have all I need to do that (update of RM or motion-gradle)? Do you see any issues with the library including things via gradle eg. Google Play Services ?

**Comment by jjaffeux** [comment]
OK I have pushed a new motion-gradle version with support for local libraries (1.2.1)

This is your project with some modifications to make it work : http://cl.ly/3x1u0O0G2k17

**Comment by buzzware** [comment]
This is great, but it isn't quite useable yet, mainly because the MotionStudioRuby/vendor/Gradle/settings.gradle you have isn't generated, and if I put it there manually, rake gradle:install blows it away and the install fails with "A problem occurred evaluating root project 'Gradle'. > Project with path ':motionstudiolib' could not be found in root project 'Gradle'."

**Comment by jjaffeux** [comment]
@buzzware  It is generated by motion-gradle, are you sure you did bundle update, and are using the last version of motion-gradle ?

**Comment by buzzware** [comment]
Its still not being generated for me

GaryMac:temp gary$ gem list --local | grep motion-gradle
motion-gradle (1.2.1, 1.1.1)
GaryMac:temp gary$ gem uninstall motion-gradle -v=1.1.1
Successfully uninstalled motion-gradle-1.1.1
GaryMac:temp gary$ git clone git@github.com:buzzware/MotionStudio.git
fatal: destination path 'MotionStudio' already exists and is not an empty directory.
GaryMac:temp gary$ rm -rf MotionStudio
GaryMac:temp gary$ git clone git@github.com:buzzware/MotionStudio.git
Cloning into 'MotionStudio'...
remote: Counting objects: 134, done.
remote: Compressing objects: 100% (72/72), done.
remote: Total 134 (delta 25), reused 124 (delta 17), pack-reused 0
Receiving objects: 100% (134/134), 79.08 KiB | 36 KiB/s, done.
Resolving deltas: 100% (25/25), done.
GaryMac:temp gary$ git log
fatal: Not a git repository (or any of the parent directories): .git
GaryMac:temp gary$ cd MotionStudio/
You are using '.rvmrc', it requires trusting, it is slower and it is not compatible with other ruby managers,
you can switch to '.ruby-version' using 'rvm rvmrc to [.]ruby-version'
or ignore this warning with 'rvm rvmrc warning ignore /Users/gary/temp/.rvmrc',
'.rvmrc' will continue to be the default project file in RVM 1 and RVM 2,
to ignore the warning for all files run 'rvm rvmrc warning ignore all.rvmrcs'.

GaryMac:MotionStudio gary$ git log
commit 33243476f8fdebc306d81fc21f2d4a816eb56d2c
Author: Gary McGhee gary@buzzware.com.au
Date:   Fri May 22 09:29:42 2015 +0800

```
updates from jjaffeux
```

commit 0827e8d6cff1d1283c980e8dc82fdcdb5a05c39d
Author: buzzware regos@buzzware.com.au
Date:   Thu May 21 23:29:21 2015 +0800

```
Create Readme
```

commit 93953e48cb8f5c4c6f940c8913c1bf320ac63625
Author: Gary McGhee gary@buzzware.com.au
Date:   Thu May 21 23:26:23 2015 +0800

```
MotionStudioRuby config
```

commit 8f3f3a0815bcba0dd870e906a912e6cffaa59ea0
Author: Gary McGhee gary@buzzware.com.au
Date:   Thu May 21 23:18:11 2015 +0800

```
app logs out test from library
```

commit 18d4a651ae28f0148cc8d992507b51fbde84a6eb
GaryMac:MotionStudio gary$ cd MotionStudioRuby
You are using '.rvmrc', it requires trusting, it is slower and it is not compatible with other ruby managers,
you can switch to '.ruby-version' using 'rvm rvmrc to [.]ruby-version'
or ignore this warning with 'rvm rvmrc warning ignore /Users/gary/temp/.rvmrc',
'.rvmrc' will continue to be the default project file in RVM 1 and RVM 2,
to ignore the warning for all files run 'rvm rvmrc warning ignore all.rvmrcs'.

GaryMac:MotionStudioRuby gary$ cat Rakefile 

# -_\- coding: utf-8 -_-

$:.unshift("/Library/RubyMotion/lib")
require 'motion/project/template/android'

begin
  require 'bundler'
  Bundler.require
rescue LoadError
end

Motion::Project::App.setup do |app|
  # Use `rake config' to see complete project settings.
  app.name = 'MotionStudioRuby'

  app.api_version = "21"
  app.development {
   app.archs << "x86" # for genymotion support
  }

  app.gradle do
    repository 'https://bintray.com/bintray/jcenter'
    repository 'https://dl.bintray.com/realm/maven'
    repository 'http://dl.bintray.com/austintaylor/gradle'
    library 'motionstudiolib', path: '/Users/gary/repos/MotionStudio/motionstudiolib'
  end

end
GaryMac:MotionStudioRuby gary$ cat Gemfile 
source 'https://rubygems.org'

gem 'rake'
gem 'motion-gradle'

# Add your dependencies here:

GaryMac:MotionStudioRuby gary$ cat Gemfile.lock
cat: Gemfile.lock: No such file or directory
GaryMac:MotionStudioRuby gary$ bundle
Fetching gem metadata from https://rubygems.org/..
Fetching version metadata from https://rubygems.org/.
Resolving dependencies...
Using rake 10.4.2
Using motion-gradle 1.2.1
Using bundler 1.9.8
Bundle complete! 2 Gemfile dependencies, 3 gems now installed.
Use `bundle show [gemname]` to see where a bundled gem is installed.
GaryMac:MotionStudioRuby gary$ nano Rakefile 
GaryMac:MotionStudioRuby gary$ pwd
/Users/gary/temp/MotionStudio/MotionStudioRuby
GaryMac:MotionStudioRuby gary$ rake gradle:install
To honour the JVM settings for this build a new JVM will be forked. Please consider using the daemon: http://gradle.org/docs/2.2.1/userguide/gradle_daemon.html.

FAILURE: Build failed with an exception.
- Where:
  Build file '/Users/gary/temp/MotionStudio/MotionStudioRuby/vendor/Gradle/build.gradle' line: 38
- What went wrong:
  A problem occurred evaluating root project 'Gradle'.
  
  > Project with path ':motionstudiolib' could not be found in root project 'Gradle'.
- Try:
  Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.

BUILD FAILED

Total time: 11.752 secs
GaryMac:MotionStudioRuby gary$ cat /Users/gary/temp/MotionStudio/MotionStudioRuby/vendor/Gradle/build.gradle
apply plugin: 'java'
apply plugin: 'eclipse'

task generateDependencies(type: Copy) {
  from sourceSets.main.runtimeClasspath
  into 'dependencies/'
}

repositories {

```
maven {
  url "/Users/gary/Applications/dev/android-rubymotion/sdk/extras/android/m2repository/"
}


maven {
  url "/Users/gary/Applications/dev/android-rubymotion/sdk/extras/google/m2repository/"
}


maven {
  url "https://bintray.com/bintray/jcenter"
}

maven {
  url "https://dl.bintray.com/realm/maven"
}

maven {
  url "http://dl.bintray.com/austintaylor/gradle"
}
```

  mavenCentral()
}

dependencies {

```
compile project(':motionstudiolib')
```

}
GaryMac:MotionStudioRuby gary$ ls /Users/gary/temp/MotionStudio/MotionStudioRuby/vendor/Gradle/
aar     build.gradle
GaryMac:MotionStudioRuby gary$ ls
Gemfile     Gemfile.lock    Rakefile    app     spec        vendor
GaryMac:MotionStudioRuby gary$ cat app/main_activity.rb 
class MainActivity < Android::App::Activity
  def onCreate(savedInstanceState)
    super
    p Example::Com::Motionstudiolib::MyModule.test
    Android::Util::Log.i("MotionStudio", Example::Com::Motionstudiolib::MyModule.test);
  end
end
GaryMac:MotionStudioRuby gary$ bundle update
Fetching gem metadata from https://rubygems.org/..
Fetching version metadata from https://rubygems.org/.
Resolving dependencies...
Using rake 10.4.2
Using motion-gradle 1.2.1
Using bundler 1.9.8
Bundle updated!
GaryMac:MotionStudioRuby gary$ rake gradle:install
To honour the JVM settings for this build a new JVM will be forked. Please consider using the daemon: http://gradle.org/docs/2.2.1/userguide/gradle_daemon.html.

FAILURE: Build failed with an exception.
- Where:
  Build file '/Users/gary/temp/MotionStudio/MotionStudioRuby/vendor/Gradle/build.gradle' line: 38
- What went wrong:
  A problem occurred evaluating root project 'Gradle'.
  
  > Project with path ':motionstudiolib' could not be found in root project 'Gradle'.
- Try:
  Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.

BUILD FAILED

Total time: 6.063 secs
GaryMac:MotionStudioRuby gary$

**Comment by jjaffeux** [comment]
Ok I found the issue... I'm stupid...

Will be fixed soon.

**Comment by jjaffeux** [comment]
Ok should be fixed by 1.3.0
